File: 347-top-k-frequent-elements/top-k-frequent-elements.java
import java.util.HashMap;
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        if (k == nums.length) {
            return nums;
        }

        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i : nums) {
            if (map.containsKey(i)) map.put(i, map.get(i) + 1);
            else map.put(i, 1);
        }
      
         // init heap 'the less frequent element first'
        Queue<Integer> heap = new PriorityQueue<>(
            (n1, n2) -> map.get(n1) - map.get(n2));

        // 2. Keep k top frequent elements in the heap
        // O(N log k) < O(N log N) time
        for (int n: map.keySet()) {
          heap.add(n);
          if (heap.size() > k) heap.poll();    
        }

        // 3. Build an output array
        // O(k log k) time
        int[] top = new int[k];
        for(int i = k - 1; i >= 0; --i) {
            top[i] = heap.poll();
        }
        return top;
        
        
    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.py

from collections import Counter
class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        c = Counter()
        for num in nums:
            if num not in c:
                c[num] = 1
            else:
                c[num] += 1
        print(c)
        print(c.most_common(k))
        topK = c.most_common(k)
        solution = []
        for item, _ in topK:
            solution.append(item)
        
        return solution


        